contracts/DomainRegistryV2.sol

```diff
@@ -0,0 +1,294 @@
+    /// @param rewardBalance - Reward balance for the domain after reward was applied
+    event RewardApplied(
+        string indexed indexedDomainName,
+        address indexed indexedDomainHolder,
+        string domainName,
+        address domainHolder,
+        uint rewardValue,
+        uint rewardBalance);
+
+    /// @notice Event that is notifying external world about applying reward to a parent node 
+    /// @param indexedDomainName - Indexed name of domain who withdraw the reward
+    /// @param indexedDomainHolder - Indexed domain holder to simplify filtering events by the holder
+    /// @param domainName - The name of domain who withdraw the reward
+    /// @param domainHolder - Domain holder of domain
+    /// @param withdrawValue - Value that was withdrawed to the domain holder
+    event RewardWithdrawed(
+        string indexed indexedDomainName,
+        address indexed indexedDomainHolder,
+        string domainName,
+        address domainHolder,
+        uint withdrawValue);
+
+
+    /// @notice Indicates that the domain name is already registered and unavailable
+    error DomainIsAlreadyRegistered();
+
+    /// @notice Indicates that the domain name was not registered
+    error DomainWasNotRegistered();
+
+    /// @notice Indicates that withdraw was failed
+    error WithdrawFailed();
+
+    /// @notice Indicates that amount of money passed is not enough to register the domain name
+    error PaymentForRegisteringDomainFailed(string message);
+
+    /// @notice Indicates that while registering sub-domain name the parent domain name doesn't exist.
+    error ParentDomainNameWasNotFound(string parentDomainName);
+
+    /// @notice Indicates that withdraw reward was failed
+    error WithdrawRewardFailed(string domainName);
+
+    /// @notice Indicates that nothing to withdraw
+    error NothingToWithdraw();
+
+
+    /// @dev Used instead of constructor due to use upgradeable contract approach
+    function initialize(address owner_, uint registrationPrice_) initializer public {
+        __Ownable_init(owner_);
+        _getRegistryStorage().registrationPrice = registrationPrice_;
+    }
+
+    receive() external payable { }
+    fallback() external payable { }
+
+    /// @notice Guarantees that only available domains can be passed to a method with the modifier
+    modifier availableDomain(string memory domain) {
+        if (isDomainRegistered(domain))
+            revert DomainIsAlreadyRegistered();
+        _;
+    }
+
+    /// @notice Guarantees that only registered domains can be passed to a method with the modifier
+    modifier onlyRegisteredDomain(string memory domain) {
+        if (!isDomainRegistered(domain))
+            revert DomainWasNotRegistered();
+        _;
+    }
+
+    /// @notice Checks if domain has been already registered
+    /// @param domainName - The name of domain to check
+    function isDomainRegistered(string memory domainName)
```

Comment:
// NatSpec structure
// view function should be declared  below the external function
// // - 1 point

Reply:
done